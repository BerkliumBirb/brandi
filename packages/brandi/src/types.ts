import { Tag } from './pointers';

type UnknownConstructor<T extends Object = Object> = new (...args: any[]) => T;
type UnknownFunction<T = unknown> = (...args: any[]) => T;

export type UnknownCreator<T = unknown> =
  | UnknownConstructor<T>
  | UnknownFunction<T>;

export type UnknownCreatorParameters<
  T extends UnknownCreator
> = T extends UnknownConstructor
  ? ConstructorParameters<T>
  : T extends UnknownFunction
  ? Parameters<T>
  : never;

export type ResolutionCondition = Tag | UnknownCreator;

export type Factory<T extends unknown, A extends unknown[] = []> = (
  ...args: A
) => T;
