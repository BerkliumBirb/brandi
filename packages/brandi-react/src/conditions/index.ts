export * from './ConditionsContext';
export * from './ConditionsProvider';
export * from './tagged';
export * from './useConditions';
