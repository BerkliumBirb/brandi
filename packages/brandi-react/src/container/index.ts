export * from './ContainerContext';
export * from './ContainerProvider';
export * from './useContainer';
