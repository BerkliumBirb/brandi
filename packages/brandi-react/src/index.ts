export { ContainerProvider } from './container';
export { createInjectionHooks, useInjection } from './injection';
export { tagged } from './tagged';
