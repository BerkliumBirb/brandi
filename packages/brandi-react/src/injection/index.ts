export * from './createInjectionHooks';
export * from './useInjection';
