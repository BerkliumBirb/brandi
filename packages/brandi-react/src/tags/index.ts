export * from './tagged';
export * from './TagsContext';
export * from './TagsProvider';
export * from './useTags';
