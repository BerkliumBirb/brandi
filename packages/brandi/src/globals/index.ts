export * from './entitiesRegistry';
export * from './injectsRegistry';
export * from './tagsRegistry';
