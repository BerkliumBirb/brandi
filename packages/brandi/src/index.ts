export { Container, createContainer } from './container';
export type {
  Tag,
  Token,
  TokenType,
  TokenTypeMap,
  TokenValue,
  RequiredToken,
  OptionalToken,
} from './pointers';
export { tag, token } from './pointers';
export { injected, tagged } from './registrators';
export type { Factory, ResolutionCondition } from './types';
