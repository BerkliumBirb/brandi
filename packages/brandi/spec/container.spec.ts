import { Container, createContainer, token } from '../src';

import { setEnv } from './utils';

describe('container', () => {
  it('throws error when the token was not bound', () => {
    const TOKENS = {
      some: token<unknown>('some'),
    };

    const container = new Container();

    expect(() => container.get(TOKENS.some)).toThrowErrorMatchingSnapshot();
  });

  it("returns an unlinked container from 'clone' method", () => {
    const parentValue = 1;
    const someValue = 2;
    const anotherValue = 3;
    const copiedValue = 4;

    const TOKENS = {
      parent: token<number>('parent'),
      original: token<number>('original'),
      copied: token<number>('copied'),
    };

    const parentContainer = new Container();
    parentContainer.bind(TOKENS.parent).toConstant(parentValue);

    const originalContainer = new Container().extend(parentContainer);
    originalContainer.bind(TOKENS.original).toConstant(someValue);

    const copiedContainer = originalContainer.clone();
    copiedContainer.bind(TOKENS.original).toConstant(anotherValue);
    copiedContainer.bind(TOKENS.copied).toConstant(copiedValue);

    expect(originalContainer.get(TOKENS.original)).toBe(someValue);
    expect(() =>
      originalContainer.get(TOKENS.copied),
    ).toThrowErrorMatchingSnapshot();

    expect(copiedContainer.get(TOKENS.parent)).toBe(parentValue);
    expect(copiedContainer.get(TOKENS.original)).toBe(anotherValue);
    expect(copiedContainer.get(TOKENS.copied)).toBe(copiedValue);
  });

  it('returns a dependency from the parent container', () => {
    const value = 1;

    const TOKENS = {
      value: token<number>('value'),
    };

    const parentContainer = new Container();
    parentContainer.bind(TOKENS.value).toConstant(value);

    const childContainer = new Container().extend(parentContainer);

    expect(childContainer.get(TOKENS.value)).toBe(value);
  });

  it('changes parent container', () => {
    const firstValue = 1;
    const secondValue = 2;

    const TOKENS = {
      value: token<number>('value'),
    };

    const firstContainer = new Container();
    firstContainer.bind(TOKENS.value).toConstant(firstValue);

    const secondContainer = new Container();
    secondContainer.bind(TOKENS.value).toConstant(secondValue);

    const container = new Container().extend(firstContainer);

    expect(container.get(TOKENS.value)).toBe(firstValue);

    container.extend(secondContainer);

    expect(container.get(TOKENS.value)).toBe(secondValue);
  });

  it('clears parent container', () => {
    const value = 1;

    const TOKENS = {
      value: token<number>('value'),
    };

    const parentContainer = new Container();
    parentContainer.bind(TOKENS.value).toConstant(value);

    const childContainer = new Container().extend(parentContainer);

    expect(childContainer.get(TOKENS.value)).toBe(value);

    childContainer.extend(null);

    expect(() =>
      childContainer.get(TOKENS.value),
    ).toThrowErrorMatchingSnapshot();
  });

  it("rebinds a parent container's binding in the child container", () => {
    const someValue = 1;
    const anotherValue = 2;

    const TOKENS = {
      value: token<number>('value'),
    };

    const parentContainer = new Container();
    parentContainer.bind(TOKENS.value).toConstant(someValue);

    const childContainer = new Container().extend(parentContainer);
    childContainer.bind(TOKENS.value).toConstant(anotherValue);

    expect(parentContainer.get(TOKENS.value)).toBe(someValue);
    expect(childContainer.get(TOKENS.value)).toBe(anotherValue);
  });

  it('captures the container bindings to a snapshot', () => {
    const someValue = 1;
    const anotherValue = 2;
    const additionalValue = 3;

    const TOKENS = {
      value: token<number>('value'),
      additional: token<number>('additional'),
    };

    const container = new Container();
    container.bind(TOKENS.value).toConstant(someValue);

    container.capture();

    container.bind(TOKENS.value).toConstant(anotherValue);
    container.bind(TOKENS.additional).toConstant(additionalValue);

    expect(container.get(TOKENS.value)).toBe(anotherValue);
    expect(container.get(TOKENS.additional)).toBe(additionalValue);

    container.restore();

    expect(container.get(TOKENS.value)).toBe(someValue);
    expect(() =>
      container.get(TOKENS.additional),
    ).toThrowErrorMatchingSnapshot();

    container.bind(TOKENS.value).toConstant(anotherValue);
    container.bind(TOKENS.additional).toConstant(additionalValue);

    expect(container.get(TOKENS.value)).toBe(anotherValue);
    expect(container.get(TOKENS.additional)).toBe(additionalValue);

    container.restore();

    expect(container.get(TOKENS.value)).toBe(someValue);
    expect(() =>
      container.get(TOKENS.additional),
    ).toThrowErrorMatchingSnapshot();
  });

  it('captures the container parent to a snapshot', () => {
    const firstValue = 1;
    const secondValue = 2;

    const TOKENS = {
      value: token<number>('value'),
    };

    const firstContainer = new Container();
    firstContainer.bind(TOKENS.value).toConstant(firstValue);

    const secondContainer = new Container();
    secondContainer.bind(TOKENS.value).toConstant(secondValue);

    const container = new Container().extend(firstContainer);

    container.capture();

    container.extend(secondContainer);

    expect(container.get(TOKENS.value)).toBe(secondValue);

    container.restore();

    expect(container.get(TOKENS.value)).toBe(firstValue);

    container.extend(secondContainer);

    expect(container.get(TOKENS.value)).toBe(secondValue);

    container.restore();

    expect(container.get(TOKENS.value)).toBe(firstValue);
  });

  it('logs an error when trying to restore a non-captured container state', () => {
    const spy = jest.spyOn(console, 'error').mockImplementation(() => null);

    const container = new Container();
    container.restore();

    expect(spy).toHaveBeenCalledTimes(1);
    expect(spy.mock.calls[0]?.[0]).toMatchSnapshot();

    spy.mockRestore();
  });

  it('does not log an error when restoring the captured container state', () => {
    const spy = jest.spyOn(console, 'error').mockImplementation(() => null);

    const container = new Container();
    container.capture();
    container.restore();

    expect(spy).toHaveBeenCalledTimes(0);

    spy.mockRestore();
  });

  it("skips the logging in 'production' env", () => {
    const restoreEnv = setEnv('production');
    const spy = jest.spyOn(console, 'error').mockImplementation(() => null);

    const container = new Container();
    container.restore();

    expect(spy).toHaveBeenCalledTimes(0);

    restoreEnv();
    spy.mockRestore();
  });

  it("creates a container by 'createContainer'", () => {
    expect(createContainer()).toBeInstanceOf(Container);
  });

  describe('typings', () => {
    it('requires to bind the same type of dependency and token', () => {
      expect.assertions(0);

      class Some {
        public some = true;
      }

      class Another {
        public another = true;
      }

      const createAnother = (): Another => new Another();

      const TOKENS = {
        num: token<number>('num'),
        some: token<Some>('some'),
      };

      const container = new Container();

      // @ts-expect-error: Argument of type 'string' is not assignable to parameter of type 'number'.
      container.bind(TOKENS.num).toConstant('');

      // @ts-expect-error: Argument of type 'typeof Another' is not assignable to parameter of type 'UnknownConstructor<Some>'.
      container.bind(TOKENS.some).toInstance(Another).inTransientScope();

      // @ts-expect-error: Argument of type '() => Another' is not assignable to parameter of type 'UnknownCreator<Some>'.
      container.bind(TOKENS.some).toInstance(createAnother).inTransientScope();
    });

    it('does not allow to bind an optional token', () => {
      expect.assertions(0);

      const TOKENS = {
        some: token<number>('some'),
      };

      const container = new Container();

      // @ts-expect-error: Argument of type 'OptionalToken<number>' is not assignable to parameter of type 'Token<unknown>'.
      container.bind(TOKENS.some.optional);
    });
  });
});
