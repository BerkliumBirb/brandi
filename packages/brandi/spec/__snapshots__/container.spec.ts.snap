// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`container captures the container bindings to a snapshot 1`] = `"No matching bindings found for 'additional' token."`;

exports[`container captures the container bindings to a snapshot 2`] = `"No matching bindings found for 'additional' token."`;

exports[`container clears parent container 1`] = `"No matching bindings found for 'value' token."`;

exports[`container logs an error when trying to restore a non-captured container state 1`] = `"Error: It looks like a trying to restore a non-captured container state. Did you forget to call 'capture()' method?"`;

exports[`container returns an unlinked container from 'clone' method 1`] = `"No matching bindings found for 'copied' token."`;

exports[`container throws error when the token was not bound 1`] = `"No matching bindings found for 'some' token."`;
